### ISSUE TITLE
Add escape-sequence to bookmark places to scroll to

### ISSUE THREAD

**Issue by pyhedgehog** [mintty/mintty#569]
This can be very useful to scroll to previous command (see Maximus5/ConEmu#787).

**Comment by mintty** [comment]
@kiwiz: Can you imagine to extend your search code (#85) with a function to scroll back so the previous line that has a certain marker (which I would introduce in struct termline) gets positioned on the screen, say in the top line? I could add the marking code so we can get a nice fancy feature here.

What would be a suitable escape sequence? Is there any terminal providing such a feature already?

**Comment by kiwiz** [comment]
Hmm, this should be simple-ish. Not sure about the esc sequence though. I don't know of any terminal emulator that offers this. I'll probably have time to make the changes this weekend.

**Comment by mintty** [comment]
Thanks a lot. I could also imagine to search for a constant, configurable string which people could include in their prompt. Invoked by Shift+left/right or Control+Shift+(Page)Up/Down.

**Comment by mintty** [comment]
This feature is now available in the repository version.
It's not yet documented; include ESC[?7711h in your prompt, then use Shift+cursor-left/right to skip to scroll marker positions. Highlighting colour is currently the same as for the scrollback search.
Feedback appreciated.

**Comment by mkq** [comment]
Great! Thank you very much. I notices some issues though (with version fffdf62):

$MINTTY_SCREEN starts at the correct scroll position, but ends at the overall last line.

$MINTTY_COMMAND used inside the prompt gives me the command I'm about to start or the currently executing if that is running long enough. So instead, I print ESC[?7711h in zsh's preexec function. That gives the last finished command's output as I wanted, except its first line is somehow always missing.

$MINTTY_SELECTION: With something selected, there's a strange side effect: Mintty's displayed selection changes to a region of full rows. I think it's from the last ESC[?7711h to the last line (regardless of whether the actual selection is in that range or somewhere nearer to the top).
What I mean by "displayed selection" is: It just looks as if I had selected those lines; what is copied to the clipboard is correct (what I actually selected).
BTW, the variable's value is correct in both situations (with or without selection).

The mintty / zsh config I have tested with:
```
UserCommands=^Atest:printf '===sel===\n%s\n===cmd===\n%s\n' "$MINTTY_SELECTION" "$MINTTY_COMMAND">/tmp/mintty.log
```
```
tputReset="%{$(tput sgr0)%}"
PS1=$(strjoin '' \
	$'\n' \
	'%{ESC[48;5;7m%}' \
	'%(0?..' '%{ESC[1m%}%{ESC[38;5;1m%}' '[status %?] ' "$tputReset" '%{ESC[48;5;9m%}' ')' \
	'$(histLockWarn)' "$tputReset" '%{ESC[48;5;7m%}' '%~' \
	'$(reset_color= git_super_status) ' \
	'%{ESC[38;5;12mESC[1m%}' '%{ESC[K%}' \
	"$tputReset" $'\n> '
)

preexec() {
	local cmd
	print -n 'ESC[?7711h'
	if [ $#1 -le 40 ]; then cmd=$1; else cmd="${1[1,37]}..."; fi
	cmd=$(perl -ple 's{([^[:graph:] ])}{sprintf("\\%03o", ord($1))}ge' <<< "$cmd")
	print -n '\e]0;'
	print -rn -- "$cmd (zsh "
	if [ ${+zshtitle} -ne 0 ]; then print -Pn "$zshtitle"; else print -Pn "$$"; fi
	print -n ")\\a"
	lastStartTime=$SECONDS
}
```
Escape characters replaced with "ESC".
`strjoin ''` just joins the args into a single line of output.
histLockWarn prints nothing if all is well.
git_super_status printed nothing in my tests, because I was not in a git directory.

**Comment by mintty** [comment]
Thanks for the feedback; these are actually comments to #100, maybe you can repeat them there, for the records, and we'll delete it here?

> $MINTTY_SCREEN starts at the correct scroll position, but ends at the overall last line.

That's its purpose, basically to provide the whole screen contents, just extend it backwards when scrolled back.

> $MINTTY_COMMAND used inside the prompt gives me the command I'm about to start or the currently executing if that is running long enough. So instead, I print ESC[?7711h in zsh's preexec function. That gives the last finished command's output as I wanted, except its first line is somehow always missing.

What do you mean by MINTTY_COMMAND inside the prompt? It's only set for user-defined commands, not for the shell.
I'm not familiar with preexec but I guess it's output appears after the command line already, so on the first command output line, which is thereby marked as "prompt line" (that's how the feature is designed, maybe I can describe that clearer). That would explain the effect.

> $MINTTY_SELECTION: With something selected, there's a strange side effect: Mintty's displayed selection changes to a region of full rows. I think it's from the last ESC[?7711h to the last line (regardless of whether the actual selection is in that range or somewhere nearer to the top).
What I mean by "displayed selection" is: It just looks as if I had selected those lines; what is copied to the clipboard is correct (what I actually selected).
BTW, the variable's value is correct in both situations (with or without selection).

The contents of MINTTY_SELECTION isn't related to the scroll markers. But it obviously affects visual highlighting (and maybe related there, I'll check) which is not intended.

> The mintty / zsh config I have tested with: ...
I hope you don't suppose me to analyse your complex prompt setup...

**Comment by mkq** [comment]
> Thanks for the feedback; these are actually comments to #100, maybe you can repeat them there, for the records, and we'll delete it here?

Sorry. I copied my comments, please feel free to delete them (and this reply) here.
I'll answer in #100 afterwards.

**Comment by mintty** [comment]
Released with 2.7.5.

**Comment by mintty** [comment]
Released 2.7.7.
